app/controllers/fhir/r4/api_controller.rb

```diff
@@ -424,6 +424,120 @@ def all
+    return if doorkeeper_authorize!(*PROVENANCE_READ_SCOPES)
+
+    status_not_acceptable_with_custom_errors(["'Accept' header must have a value of 'application/fhir+json'"]) && return unless accept_header?
+
+    unless request.headers['Prefer']&.include?('respond-async')
+      status_unprocessable_entity_with_custom_errors(["'Prefer' header must have value of 'respond-async'"],
+                                                     '') && return
+    end
+
+    unless @m2m_workflow
+      status_unauthorized_with_custom_errors(['Bulk export requires a client application registered for the Backend Services Workflow']) && return
+    end
+
+    client_app = current_client_application
+    if client_app.exported_recently?
+      status_too_many_requests_with_custom_errors(['Client already initiated an export of this type in the last 15 minutes. Please try again later']) && return
+    end
+
+    begin
+      since = params.permit(:_since)[:_since]
+      since = DateTime.parse(since) unless since.nil?
+    rescue Date::Error
+      status_unprocessable_entity_with_custom_errors(['Invalid Date in _since parameter. Please use the FHIR instant datatype'], '') && return
+    end
+
+    # Create the download to uniquely id this request, and queue the ExportFhirJob
+    download = ApiDownload.create(application_id: client_app.id, url: request.url)
+    export_job = ExportFhirJob.perform_later(client_app, download, { since: since })
+    # Add the job_id of the job to the download for reference in status requests
+    download.update(job_id: export_job.provider_job_id)
+
+    respond_to do |format|
+      format.any { head(:accepted, content_location: "#{root_url}fhir/r4/ExportStatus/#{download.id}") }
+    end
+  end
+
+  # Give an update on the status of a bulk FHIR data request
+  #
+  # GET /fhir/r4/ExportStatus/[:id]
+  def export_status
+    id = params.permit(:id)[:id]
+    download = ApiDownload.find_by_id(id)
+
+    case Sidekiq::Status.status(download&.job_id)
+    when :complete
+      response_json = {
+        transactionTime: Sidekiq::Status.get(download.job_id, :transaction_time),
+        request: download.url,
+        requiresAccessToken: true,
+        output: ActiveStorage::Blob.where(id: download.files.pluck(:blob_id)).pluck(:filename).map do |file|
+          {
+            type: file.split('.').first,
+            url: "#{root_url}fhir/r4/ExportFiles/#{id}/#{file}"
+          }
+        end
+      }
+      render json: response_json.to_json, status: :ok
+    when :failed
+      status_server_error(['Export failed'])
+    when :working, :queued, :retrying
+      head :accepted
+    else
+      status_not_found_with_custom_errors(['No export found for this ID'])
+    end
+  end
+
+  # Return an bulk exported ndjson file
+  #
+  # GET /ExportFiles/[:id]/[:filename]
+  def export_files
+    id = params.permit(:id)[:id]
+    filename = params.permit(:filename)[:filename]
+    resource_type = filename&.split('.')&.first&.downcase
```

Comment:
According to the specification, we control the URL that is accessed. [Link](https://hl7.org/fhir/uv/bulkdata/export/index.html#endpoint-2)

Are we keeping to some convention where we make a user request a "filename"? 
Can one simply include the _type and we can save the parsing? Essentially filename => resource_type and we can just set the file name as part of the response based on the type. This may also remove the constraints within `routes.rb`

Reply:
Yeah, I was just conforming to the example the spec gives where they include the `.ndjson`. But I agree, it's easier for us and for users if we just do `resource_type`.

Reply:
Done in https://github.com/SaraAlert/SaraAlert/pull/987/commits/6977cd314b84a7c5ecfd8d2d2ac200bf5d218e7b.